### PR TITLE
Add runtime field section to Field Capabilities API

### LINKED ISSUE #68117: Add a runtime field section to the Field Capabilities API
[Runtime fields defined in a search request](https://www.elastic.co/guide/en/elasticsearch/reference/master/runtime-search-request.html) will not appear in the output of the field capabilities API. Merging these runtime fields in the output of the field capabilities API is not trivial since we have some internal logic to decide the capabilities of fields based on their mappings.
In order to make it easier for Kibana, QL, ml or any consumer of runtime fields, the field capabilities API should be able to handle a runtime section just like search requests. The section would be used internally as an override in the mappings of all the indices that the request targets:

```yml
GET my_data_stream/_field_caps?fields=*
{
  "runtime_mappings": {
    "day_of_week": {
      "type": "keyword",
      "script": {
        "source": "emit(doc['@timestamp'].value.dayOfWeekEnum.getDisplayName(TextStyle.FULL, Locale.ROOT))"
      }
    }
  }
}
```

Would return all the fields present in the mapping as well as the runtime fields defined in the request (as an override if the field already exists in the index mappings).

Relates #68115

### PR DESCRIPTION
**Pull request by cbuescher**
Currently runtime fields from search requests don't appear in the output of the
field capabilities API, but some consumer of runtime fields would like to see
runtime section just like they are defined in search requests reflected and
merged into the field capabilities output.
This change adds parsing of a "runtime_mappings" section equivallent to the one
on search requests to the `_field_caps` endpoint, passes this section down to
the shard level where any runtime fields defined here overwrite the mapping of
the targetet indices.

Closes #68117

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesRequestTests.java [hunk 1]
```diff
@@ -67,6 +87,11 @@ protected FieldCapabilitiesRequest mutateInstance(FieldCapabilitiesRequest insta
         });
         mutators.add(request -> request.setMergeResults(request.isMergeResults() == false));
         mutators.add(request -> request.includeUnmapped(request.includeUnmapped() == false));
+        mutators.add(request -> request.nowInMillis(request.nowInMillis() != null ? request.nowInMillis() + 1 : 1L));
+        mutators.add(
+            request -> request.indexFilter(request.indexFilter() != null ? request.indexFilter().boost(2) : QueryBuilders.matchAllQuery())
+        );
+        mutators.add(request -> request.runtimeFields(Collections.singletonMap("other_key", "other_value")));
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesRequestTests.java:95]: would it make sense to add a test for the toXContent output of the request if we don't test it yet?

**Comment by cbuescher** [comment]
One caveat of this PR: I found that we do not propagate back exceptions thrown in the shard operation in `TransportFieldCapabilitiesIndexAction` to the calling nodes in a good way currently. If there are errors in the `runtime_mappins` section, e.g.
```
{
  "runtime_mappings": {
    "day_of_week": {
      "type": "some_bad_type",
      "script": {
        "source": "emit(doc['@timestamp'].value.dayOfWeekEnum.getDisplayName(TextStyle.FULL, Locale.ROOT))"
      }
    }
  }
}
```
The resulting MapperParsingException in that is thrown when parsing the runtime fields upon creating the `SearchExecutionContext` on the shard is dropped somewhere. Fixing this is a bit more involved so I would prefer doing this in a follow up.

File: docs/reference/search/field-caps.asciidoc [hunk 2]
```diff
@@ -88,6 +88,13 @@ include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=index-ignore-unavailab
 (Optional,  <<query-dsl,query object>> Allows to filter indices if the provided
 query rewrites to `match_none` on every shard.
 
+`runtime_mappings`::
+(Optional, object)
+Defines ad-hoc <<runtime-search-request,runtime fields>> in the request similar
+to the way it is done in <<search-api-body-runtime, search requests>>. This fields
```
Comment (jimczi) [inline docs/reference/search/field-caps.asciidoc:94]:
```suggestion
to the way it is done in <<search-api-body-runtime, search requests>>. These fields
```

**Review by jimczi — APPROVED** [review]
LGTM

**Review by javanna — APPROVED** [review]
The change looks good to me, error handling looks like wasn't part of the initial design of field_caps and it now becomes more important as scripts may not compile etc. Although field_caps should not be used to validate scripts (and possibly we should not even need to compile them for field_caps), returning an empty response whenever there is an error does not seem like a good idea.

I'd say merge this one and let's discuss and work on the error handling as a follow-up?

File: server/src/test/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesRequestTests.java (hunk 1 above)
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesRequestTests.java:95]: sure, will do

**Pull request merged**